
Pangenome analysis
==================

For the pangenome analysis we used \ `orthofinder <https://github.com/davidemms/OrthoFinder>`_\  v2.2.1. All analysis used the output of this tools.

Phylogenetic
------------

For product the phylogenetic tree we used two script (OG_phylo.py & alnToSnp.py) then we used \ `RaXML <https://github.com/stamatak/standard-RAxML>`_\  v8.2.4.

OG_phylo
^^^^^^^^

This program is used to search for orthologues groupe  of core genome single copy and create a consensus for phylogenetic tree. For create a consensus, this script align and clean each orthogroups select with translatorX.

Mandatory installation
~~~~~~~~~~~~~~~~~~~~~~


- \ `Python >=3.7 <https://www.python.org/downloads/>`_\ 

- \ `translatorX == 1.1 <http://translatorx.co.uk/>`_\ 

Arguments take by OG_phylo
~~~~~~~~~~~~~~~~~~~~~~~~~~


- **-g, --group** (type : string) : path of the result of orthofinder (format txt)
- **-c, --count** (type : string) :  path of the count result of orthofinder (format csv)
- **-f, --fasta** (type : string) : path of the fasta which contains all sequence of all strain used for orthofinder
- **-o, --outdir** (type : none) : path of output directory for all output file

**Exemple:**

.. code-block::

   cat /homedir/user/work/directory_input_orthofinder/*.fasta > /homedir/user/work/all_sequence.fasta
   OG_phylo.py -g /homedir/user/work/output/Orthogroups.txt -c /homedir/user/work/output/Orthogroups.GeneCount.csv -f /homedir/user/work/all_sequence.fasta -o /homedir/user/work/output/phylo/

.. raw:: html

	<br />


AlnToSnp
^^^^^^^^

This Programme is used to keep only SNP from a alignement

Mandatory installation
~~~~~~~~~~~~~~~~~~~~~~


- \ `Python >=3.7 <https://www.python.org/downloads/>`_\

Arguments take by alnToSnp
~~~~~~~~~~~~~~~~~~~~~~~~~~


- **-a, --alignement** (type : string) :  path of the count result of orthofinder (format csv)
- **-t, --type** (type : string) : Alignement type (ex : fasta)
- **-o, --output** (type : none) : path of output file

**Exemple:**

.. code-block::

   OG_phylo.py -a /homedir/user/work/output/phylo/sequence_merge.fasta -t fasta -o /homedir/user/work/output/sequence_aln.fasta

Correction_blast
----------------

For each orthogroup, the script checks if all isolates absent of orthogroup isn't a false absence. If this isn't the case, the script corrects the gene.count file generated by orthofinder.

Mandatory installation
~~~~~~~~~~~~~~~~~~~~~~


- \ `Python >=3.7 <https://www.python.org/downloads/>`_\ 

Arguments take by Correction_blast
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


- **-d, --directory** (type : string) : path of the directory which contains all OG fasta file
- **-c, --count** (type : string) : path of the count result of orthofinder (format csv)
- **-db, --database** (type : string) : path of the all assembly merge file with scafold name : Souche_Scaffold_1
- **-o, --outdir** (type : string) :  path of output directory for all output file

**Exemple:**

.. code-block::

   OG_phylo.py -g /homedir/user/work/OG_fasta/ -c /homedir/user/work/output/Orthogroups.GeneCount.csv -db /homedir/user/work/all_scaffold.fasta -o /homedir/user/work/output/phylo/
