

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Pangenome analysis &mdash; Maize paper script</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Annotation pipeline" href="Annotation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Maize_paper
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Assembly Script</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Assembly.html">Assembly Script</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Assembly.html#abyss-launch">ABYSS_launch</a></li>
<li class="toctree-l2"><a class="reference internal" href="Assembly.html#formatfastaname">formatFastaName</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Secretome Selection</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Secretome.html">Secretome Selection</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Secretome.html#secretome-pipeline">Secretome_Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="Secretome.html#comparaison-secretome">Comparaison_secretome</a></li>
<li class="toctree-l2"><a class="reference internal" href="Secretome.html#eliminateremotif">EliminateREmotif</a></li>
<li class="toctree-l2"><a class="reference internal" href="Secretome.html#selectiontmhmm">SelectionTMHMM</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Annotation pipeline</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Annotation.html">Annotation pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Annotation.html#mandatory-installation">Mandatory installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="Annotation.html#prepare-config-file">Prepare config file</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Pangenome analysis</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pangenome analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#phylogenetic">Phylogenetic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#correction-blast">Correction_blast</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Maize_paper</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Pangenome analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/Pangenome.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pangenome-analysis">
<h1>Pangenome analysis<a class="headerlink" href="#pangenome-analysis" title="Permalink to this headline">¶</a></h1>
<p>For the pangenome analysis we used <a class="reference external" href="https://github.com/davidemms/OrthoFinder">orthofinder</a> v2.2.1. All analysis used the output of this tools.</p>
<div class="section" id="phylogenetic">
<h2>Phylogenetic<a class="headerlink" href="#phylogenetic" title="Permalink to this headline">¶</a></h2>
<p>For product the phylogenetic tree we used two script (OG_phylo.py &amp; alnToSnp.py) then we used <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RaXML</a> v8.2.4.</p>
<div class="section" id="og-phylo">
<h3>OG_phylo<a class="headerlink" href="#og-phylo" title="Permalink to this headline">¶</a></h3>
<p>This program is used to search for orthologues groupe  of core genome single copy and create a consensus for phylogenetic tree. For create a consensus, this script align and clean each orthogroups select with translatorX.</p>
<div class="section" id="mandatory-installation">
<h4>Mandatory installation<a class="headerlink" href="#mandatory-installation" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
<li><p><a class="reference external" href="http://translatorx.co.uk/">translatorX == 1.1</a></p></li>
</ul>
</div>
<div class="section" id="arguments-take-by-og-phylo">
<h4>Arguments take by OG_phylo<a class="headerlink" href="#arguments-take-by-og-phylo" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><strong>-g, –group</strong> (type : string) : path of the result of orthofinder (format txt)</p></li>
<li><p><strong>-c, –count</strong> (type : string) :  path of the count result of orthofinder (format csv)</p></li>
<li><p><strong>-f, –fasta</strong> (type : string) : path of the fasta which contains all sequence of all strain used for orthofinder</p></li>
<li><p><strong>-o, –outdir</strong> (type : none) : path of output directory for all output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">directory_input_orthofinder</span><span class="o">/*.</span><span class="n">fasta</span> <span class="o">&gt;</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">all_sequence</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">OG_phylo</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">g</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">Orthogroups</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">Orthogroups</span><span class="o">.</span><span class="n">GeneCount</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">f</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">all_sequence</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phylo</span><span class="o">/</span>
</pre></div>
</div>
<br /></div>
</div>
<div class="section" id="alntosnp">
<h3>AlnToSnp<a class="headerlink" href="#alntosnp" title="Permalink to this headline">¶</a></h3>
<p>This Programme is used to keep only SNP from a alignement</p>
<div class="section" id="id1">
<h4>Mandatory installation<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
</ul>
</div>
<div class="section" id="arguments-take-by-alntosnp">
<h4>Arguments take by alnToSnp<a class="headerlink" href="#arguments-take-by-alntosnp" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><strong>-a, –alignement</strong> (type : string) :  path of the count result of orthofinder (format csv)</p></li>
<li><p><strong>-t, –type</strong> (type : string) : Alignement type (ex : fasta)</p></li>
<li><p><strong>-o, –output</strong> (type : none) : path of output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">OG_phylo</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">a</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phylo</span><span class="o">/</span><span class="n">sequence_merge</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">t</span> <span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">sequence_aln</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="correction-blast">
<h2>Correction_blast<a class="headerlink" href="#correction-blast" title="Permalink to this headline">¶</a></h2>
<p>For each orthogroup, the script checks if all isolates absent of orthogroup isn’t a false absence. If this isn’t the case, the script corrects the gene.count file generated by orthofinder.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
</ul>
<ul class="simple">
<li><p><strong>-d, –directory</strong> (type : string) : path of the directory which contains all OG fasta file</p></li>
<li><p><strong>-c, –count</strong> (type : string) : path of the count result of orthofinder (format csv)</p></li>
<li><p><strong>-db, –database</strong> (type : string) : path of the all assembly merge file with scafold name : Souche_Scaffold_1</p></li>
<li><p><strong>-o, –outdir</strong> (type : string) :  path of output directory for all output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">OG_phylo</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">g</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">OG_fasta</span><span class="o">/</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">Orthogroups</span><span class="o">.</span><span class="n">GeneCount</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">db</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">all_scaffold</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phylo</span><span class="o">/</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="Annotation.html" class="btn btn-neutral float-left" title="Annotation pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2019-2020, Florian CHARRIAT (INRAE)

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXXXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>