

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Secretome Selection &mdash; Maize paper script</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Annotation pipeline" href="Annotation.html" />
    <link rel="prev" title="Assembly Script" href="Assembly.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Maize_paper
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Assembly Script</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Assembly.html">Assembly Script</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Assembly.html#abyss-launch">ABYSS_launch</a></li>
<li class="toctree-l2"><a class="reference internal" href="Assembly.html#formatfastaname">formatFastaName</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Secretome Selection</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Secretome Selection</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#secretome-pipeline">Secretome_Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#comparaison-secretome">Comparaison_secretome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#eliminateremotif">EliminateREmotif</a></li>
<li class="toctree-l2"><a class="reference internal" href="#selectiontmhmm">SelectionTMHMM</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Annotation pipeline</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Annotation.html">Annotation pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Annotation.html#mandatory-installation">Mandatory installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="Annotation.html#prepare-config-file">Prepare config file</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Pangenome analysis</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Pangenome.html">Pangenome analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Pangenome.html#phylogenetic">Phylogenetic</a></li>
<li class="toctree-l2"><a class="reference internal" href="Pangenome.html#correction-blast">Correction_blast</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Maize_paper</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Secretome Selection</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/Secretome.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="secretome-selection">
<h1>Secretome Selection<a class="headerlink" href="#secretome-selection" title="Permalink to this headline">¶</a></h1>
<p>For select secretome from protein fasta of many sample, we use secretome_Piepline. This script is used on <a class="reference external" href="https://bioinfo-agap.cirad.fr/">cirad HPC</a>, so the environment (module load command) and sge submission works only on the cirad cluster. This script does not directly launch ABySS but creates output directory and bash scripts to launch all assembly in parallel.</p>
<div class="section" id="secretome-pipeline">
<h2>Secretome_Pipeline<a class="headerlink" href="#secretome-pipeline" title="Permalink to this headline">¶</a></h2>
<p>This program is used to predict secretome. This program uses :</p>
<ol class="arabic simple">
<li><p>SignalP, TagetP, Phobius for detect the presence of signal peptide cleavage sites.</p></li>
<li><p>ComparaisonSecretome script to retrieve and compare information from the secretome prediction tools.</p></li>
<li><p>Selection_TMHMM script to select, from TMHMM ouput, only protein with no transmembrane domain or only one
transmembrane in 40 first aa.</p></li>
<li><p>EliminateREmotif script to eliminate, from ps_scan ouput, the protein with a RE retention motif.</p></li>
</ol>
<dl class="simple">
<dt>All step hasn’t launch by the script, at the end of the script a message give you a command line to enter in your</dt><dd><p>terminal for launch all job</p>
</dd>
</dl>
<div class="section" id="mandatory-installation">
<h3>Mandatory installation<a class="headerlink" href="#mandatory-installation" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="http://www.cbs.dtu.dk/services/SignalP-4.1/">SignalP == 4.1</a></p></li>
<li><p><a class="reference external" href="http://www.cbs.dtu.dk/services/TargetP/">targetP</a></p></li>
<li><p><a class="reference external" href="http://phobius.sbc.su.se/data.html">phobius</a></p></li>
<li><p><a class="reference external" href="https://prosite.expasy.org/scanprosite/">PS-scan</a></p></li>
<li><p><a class="reference external" href="http://www.cbs.dtu.dk/services/TMHMM/">TMHMM</a></p></li>
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
</ul>
</div>
<div class="section" id="arguments-take-by-secretome-pipeline">
<h3>Arguments take by secretome_Pipeline<a class="headerlink" href="#arguments-take-by-secretome-pipeline" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>-f, –file</strong> (type : string) : path of fasta files that contains all the protein</p></li>
<li><p><strong>-o, –outdir</strong> (type : string) :  path of output directory for all output file (script, log, stat and assembly)</p></li>
<li><p><strong>-p, –prosite</strong> (type : string) : path of prosite.dat file. You can upload the file <strong>ftp://ftp.expasy.org/databases/prosite/prosite.dat</strong></p></li>
<li><p><strong>-fo, –force</strong> (type : none) : Force the script to remove output data</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">secretome_Pipeline</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">d</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span> <span class="o">-</span><span class="n">p</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">prosite</span><span class="o">.</span><span class="n">dat</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="comparaison-secretome">
<h2>Comparaison_secretome<a class="headerlink" href="#comparaison-secretome" title="Permalink to this headline">¶</a></h2>
<p>This program is used to retrieve and compare information from the output of secretome prediction tools (signalP, targetP and Phobius) and create a summary table. This program is used by secretome_Pipeline.</p>
<div class="section" id="id1">
<h3>Mandatory installation<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
</ul>
</div>
<div class="section" id="arguments-take-by-comparaison-secretome">
<h3>Arguments take by comparaison_secretome<a class="headerlink" href="#arguments-take-by-comparaison-secretome" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>-s, –signalp</strong> (type : string) : path of the signalP output file</p></li>
<li><p><strong>-t, –targetp</strong> (type : string) : path of the targetp output file</p></li>
<li><p><strong>-p, –phobius</strong> (type : string) : path of the phobius output file</p></li>
<li><p><strong>-f, –fasta</strong> (type : string) : path of the input fasta file used in singalP, targetp and phobius</p></li>
<li><p><strong>-r, –rank</strong> (type : int) :  rank mini for selection by default rank = 3.</p>
<ul>
<li><p>rank 1 = protein secreted detect by all tools</p></li>
<li><p>rank 2 = protein secreted detect by two tools,</p></li>
<li><p>rank 3 : protein secreted detect by one tool</p></li>
</ul>
</li>
<li><p><strong>-o, –outdir</strong> (type : string) :  path of output directory for all output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Comparaison between all tools</span>
<span class="n">comparaisonSecretome</span><span class="o">.</span><span class="n">py</span>  <span class="o">--</span><span class="n">signalp</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_signalp</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">targetp</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_targetp</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">phobius</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_phobius</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">f</span>  <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">protein</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span>

<span class="c1"># Comparaison and selection with two tools</span>
<span class="n">comparaisonSecretome</span><span class="o">.</span><span class="n">py</span>  <span class="o">--</span><span class="n">signalp</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_signalp</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">targetp</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_targetp</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">f</span>  <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">protein</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span>

<span class="c1">#Comparaison and selection with only one tools</span>
<span class="n">comparaisonSecretome</span><span class="o">.</span><span class="n">py</span>  <span class="o">--</span><span class="n">signalp</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">output_signalp</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">f</span>  <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">protein</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="eliminateremotif">
<h2>EliminateREmotif<a class="headerlink" href="#eliminateremotif" title="Permalink to this headline">¶</a></h2>
<p>This program is used to eliminate sequence from the result of PS-scan with PS00014 pattern (RE target Motif). This program is used by secretome_Pipeline.</p>
<div class="section" id="id3">
<h3>Mandatory installation<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://www.python.org/downloads/">Python &gt;=3.7</a></p></li>
</ul>
</div>
<div class="section" id="arguments-take-by-eliminateremotif">
<h3>Arguments take by eliminateREmotif<a class="headerlink" href="#arguments-take-by-eliminateremotif" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>-p, –ps_scan</strong> (type : string) : path of the ps_scan output file</p></li>
<li><p><strong>-f, –fasta</strong> (type : string) : path of the fasta file which ps_scan has been proceed</p></li>
<li><p><strong>-o, –outdir</strong> (type : string) :  path of output directory for all output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">eliminateREmotif</span><span class="o">.</span><span class="n">py</span>  <span class="o">-</span><span class="n">p</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">TMHMM_result</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span> <span class="o">-</span><span class="n">f</span>  <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span><span class="n">protein_filter</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="selectiontmhmm">
<h2>SelectionTMHMM<a class="headerlink" href="#selectiontmhmm" title="Permalink to this headline">¶</a></h2>
<p>This program is used to selected the proteins having no transmembrane domain or only one in the 40 first amino acid. This program is used by secretome_Pipeline.</p>
<div class="section" id="arguments-take-by-selectiontmhmm">
<h3>Arguments take by selectionTMHMM<a class="headerlink" href="#arguments-take-by-selectiontmhmm" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>-t, –TMHMM</strong> (type : string) : path of the TMHMM output file</p></li>
<li><p><strong>-f, –fasta</strong> (type : string) : path of the fasta file which TMHMM has been proceed</p></li>
<li><p><strong>-o, –outdir</strong> (type : string) :  path of output directory for all output file</p></li>
</ul>
<p><strong>Exemple:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">selection_TMHMM</span><span class="o">.</span><span class="n">py</span>  <span class="o">-</span><span class="n">t</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">TMHMM_result</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span> <span class="o">-</span><span class="n">f</span>  <span class="o">/</span><span class="n">homedir</span><span class="o">/</span><span class="n">user</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">result</span><span class="o">/</span><span class="n">protein_filter</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Annotation.html" class="btn btn-neutral float-right" title="Annotation pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Assembly.html" class="btn btn-neutral float-left" title="Assembly Script" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2019-2020, Florian CHARRIAT (INRAE)

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXXXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>